python/glow/wgr/linear_model/logistic_model.py

```diff
@@ -0,0 +1,238 @@
+
+        Args:
+            blockdf : Spark DataFrame representing the beginning block matrix X
+            labeldf : Pandas DataFrame containing the target labels used in fitting the ridge models
+            sample_blocks : Dict containing a mapping of sample_block ID to a list of corresponding sample IDs
+            covdf : Pandas DataFrame containing covariates to be included in every model in the stacking
+                ensemble (optional).
+
+        Returns:
+            Two Spark DataFrames, one containing the model resulting from the fitting routine and one containing the
+            results of the cross validation procedure.
+        """
+        validate_inputs(labeldf, covdf, 'binary')
+        map_key_pattern = ['sample_block', 'label', 'alpha_name']
+        reduce_key_pattern = ['header_block', 'header', 'label', 'alpha_name']
+        model_key_pattern = ['sample_block', 'label', 'alpha_name']
+        score_key_pattern = ['sample_block', 'label']
+        p0_dict = {k: v for k, v in zip(labeldf.columns, labeldf.sum(axis=0) / labeldf.shape[0])}
+        metric = 'log_loss'
+
+        if not self.alphas:
+            self.alphas = generate_alphas(blockdf)
+
+        map_udf = pandas_udf(
+            lambda key, pdf: map_irls_eqn(key, map_key_pattern, pdf, labeldf, sample_blocks, covdf,
+                                          p0_dict, self.alphas), irls_eqn_struct,
+            PandasUDFType.GROUPED_MAP)
+
+        reduce_udf = pandas_udf(lambda key, pdf: reduce_irls_eqn(key, reduce_key_pattern, pdf),
+                                irls_eqn_struct, PandasUDFType.GROUPED_MAP)
+
+        model_udf = pandas_udf(
+            lambda key, pdf: solve_irls_eqn(key, model_key_pattern, pdf, labeldf, self.alphas),
+            model_struct, PandasUDFType.GROUPED_MAP)
+
+        score_udf = pandas_udf(
+            lambda key, pdf: score_models(key, score_key_pattern, pdf, labeldf, sample_blocks, self.
+                                          alphas, covdf, metric), cv_struct,
+            PandasUDFType.GROUPED_MAP)
+
+        modeldf = blockdf.drop('alpha') \
+            .withColumn('alpha_name', f.explode(f.array([f.lit(n) for n in self.alphas.keys()]))) \
+            .groupBy(map_key_pattern) \
+            .apply(map_udf) \
+            .groupBy(reduce_key_pattern) \
+            .apply(reduce_udf) \
+            .groupBy(model_key_pattern) \
+            .apply(model_udf) \
+            .withColumn('alpha_label_coef', f.expr('struct(alphas[0] AS alpha, labels[0] AS label, coefficients[0] AS coefficient)')) \
+            .groupBy('header_block', 'sample_block', 'header', 'sort_key', f.col('alpha_label_coef.label')) \
+            .agg(f.sort_array(f.collect_list('alpha_label_coef')).alias('alphas_labels_coefs')) \
+            .selectExpr('*', 'alphas_labels_coefs.alpha AS alphas', 'alphas_labels_coefs.label AS labels', 'alphas_labels_coefs.coefficient AS coefficients') \
```

Comment:
You might find [expand_struct](https://glow.readthedocs.io/en/latest/api-docs/pyspark-functions.html#glow.functions.expand_struct) useful here.

Reply:
I tried it but `alphas_labels_coefs` is actually an array of structs so it didn't seem to work.  *-expanding an array of structs also doesn't work I don't think, but .-selecting a field from an array of structs does work to get an array of just that field.

Reply:
Makes sense. Perhaps we will expand expand_struct to cover this case as well in the future.